In [1]:
import os
os.chdir('/Users/amira/Desktop/BIO392/exercises/day8/notebook')

In [2]:
!mkdir -p ./outputB/FastQC_Out
!fastqc -f fastq -t 2 -o ./outputB/FastQC_Out ./data/FASTQ_Chr14/B_1.fq.gz ./data/FASTQ_Chr14/B_2.fq.gz


application/gzip
application/gzip
Started analysis of B_1.fq.gz
Approx 5% complete for B_1.fq.gz
Started analysis of B_2.fq.gz
Approx 10% complete for B_1.fq.gz
Approx 5% complete for B_2.fq.gz
Approx 15% complete for B_1.fq.gz
Approx 10% complete for B_2.fq.gz
Approx 20% complete for B_1.fq.gz
Approx 15% complete for B_2.fq.gz
Approx 25% complete for B_1.fq.gz
Approx 20% complete for B_2.fq.gz
Approx 30% complete for B_1.fq.gz
Approx 25% complete for B_2.fq.gz
Approx 35% complete for B_1.fq.gz
Approx 30% complete for B_2.fq.gz
Approx 40% complete for B_1.fq.gz
Approx 35% complete for B_2.fq.gz
Approx 45% complete for B_1.fq.gz
Approx 40% complete for B_2.fq.gz
Approx 50% complete for B_1.fq.gz
Approx 45% complete for B_2.fq.gz
Approx 55% complete for B_1.fq.gz
Approx 50% complete for B_2.fq.gz
Approx 60% complete for B_1.fq.gz
Approx 55% complete for B_2.fq.gz
Approx 65% complete for B_1.fq.gz
Approx 60% complete for B_2.fq.gz
Approx 70% complete for B_1.fq.gz
Approx 65% complete for 

In [3]:
!mkdir -p ./outputB/fastp

In [4]:
!fastp -i ./data/FASTQ_Chr14/B_1.fq.gz \
    -I ./data/FASTQ_Chr14/B_2.fq.gz \
    -o ./outputB/fastp/clean_B_1.fq.gz \
    -O ./outputB/fastp/clean_B_2.fq.gz \
    -q 30 \
    -h ./outputB/fastp/B_fastp_report.html

Read1 before filtering:
total reads: 397602
total bases: 40157802
Q20 bases: 39898011(99.3531%)
Q30 bases: 38738731(96.4663%)
Q40 bases: 14612471(36.3876%)

Read2 before filtering:
total reads: 397602
total bases: 40157802
Q20 bases: 39623778(98.6702%)
Q30 bases: 38081683(94.8301%)
Q40 bases: 14121679(35.1655%)

Read1 after filtering:
total reads: 390677
total bases: 39449142
Q20 bases: 39248680(99.4918%)
Q30 bases: 38253784(96.9699%)
Q40 bases: 14566090(36.9237%)

Read2 after filtering:
total reads: 390677
total bases: 39449142
Q20 bases: 38987294(98.8293%)
Q30 bases: 37629533(95.3875%)
Q40 bases: 14086299(35.7075%)

Filtering result:
reads passed filter: 781354
reads failed due to low quality: 13270
reads failed due to too many N: 580
reads failed due to too short: 0
reads failed due to adapter dimer: 0
reads with adapter trimmed: 1190
bases trimmed due to adapters: 18554

Duplication rate: 0.000251508%

Insert size peak (evaluated by paired-end reads): 165

JSON report: fastp.json
H

In [5]:
!bwa mem -M -t 2 \
    -R "@RG\tID:A\tPL:ILLUMINA\tLB:lib1\tSM:A" \
    ./reference/chr14.fa \
    ./outputB/fastp/clean_B_1.fq.gz \
    ./outputB/fastp/clean_B_2.fq.gz | samtools sort -@ 2 -o ./outputB/bwa/B_sorted.bam -

[M::bwa_idx_load_from_disk] read 0 ALT contigs
[M::process] read 198056 sequences (20000006 bp)...
[M::process] read 198068 sequences (20000018 bp)...
[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (3, 94650, 1, 0)
[M::mem_pestat] skip orientation FF as there are not enough pairs
[M::mem_pestat] analyzing insert size distribution for orientation FR...
[M::mem_pestat] (25, 50, 75) percentile: (159, 186, 223)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (31, 351)
[M::mem_pestat] mean and std.dev: (193.85, 47.45)
[M::mem_pestat] low and high boundaries for proper pairs: (1, 415)
[M::mem_pestat] skip orientation RF as there are not enough pairs
[M::mem_pestat] skip orientation RR as there are not enough pairs
[M::mem_process_seqs] Processed 198056 reads in 14.447 CPU sec, 7.681 real sec
[M::process] read 198066 sequences (20000096 bp)...
[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (3, 94783, 2, 1)
[M::mem_pestat] skip orientatio

In [6]:
!gatk --java-options "-Xmx4G" MarkDuplicates \
        -I ./outputB/bwa/B_sorted.bam \
        -O ./outputB/bwa/B_sort_markdup.bam \
        -M ./outputB/bwa/B_sort_markdup_metrics.txt \
        --TMP_DIR .

Using GATK jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar
Running:
    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -Xmx4G -jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar MarkDuplicates -I ./outputB/bwa/B_sorted.bam -O ./outputB/bwa/B_sort_markdup.bam -M ./outputB/bwa/B_sort_markdup_metrics.txt --TMP_DIR .
18:00:39.103 INFO  NativeLibraryLoader - Loading libgkl_compression.dylib from jar:file:/opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar!/com/intel/gkl/native/libgkl_compression.dylib
[Mon May 04 18:00:40 CEST 2026] MarkDuplicates --INPUT ./outputB/bwa/B_sorted.bam --OUTPUT ./outputB/bwa/B_sort_markdup.bam --METRICS_FILE ./outputB/bwa/B_sort_markdup_metrics.txt --TMP_DIR . --MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP 50000 --MAX_FILE_HANDLES_FOR_READ_ENDS_MAP 8000 

In [7]:
!gatk --java-options "-Xmx4G" CollectWgsMetrics \
        -MQ 20 -Q 20 \
        -R ./reference/chr14.fa \
        -O ./outputB/bwa/B_sort_markdup_WGSmetrics.txt \
        -I ./outputB/bwa/B_sort_markdup.bam

Using GATK jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar
Running:
    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -Xmx4G -jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar CollectWgsMetrics -MQ 20 -Q 20 -R ./reference/chr14.fa -O ./outputB/bwa/B_sort_markdup_WGSmetrics.txt -I ./outputB/bwa/B_sort_markdup.bam
18:01:40.810 INFO  NativeLibraryLoader - Loading libgkl_compression.dylib from jar:file:/opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar!/com/intel/gkl/native/libgkl_compression.dylib
[Mon May 04 18:01:41 CEST 2026] CollectWgsMetrics --INPUT ./outputB/bwa/B_sort_markdup.bam --OUTPUT ./outputB/bwa/B_sort_markdup_WGSmetrics.txt --MINIMUM_MAPPING_QUALITY 20 --MINIMUM_BASE_QUALITY 20 --REFERENCE_SEQUENCE ./reference/chr14.fa --COVERAGE_CAP 250 --LOCUS_ACCUMULATI

In [8]:
!samtools faidx ./reference/chr14.fa
!gatk --java-options "-Xmx4G" CreateSequenceDictionary \
    -R ./reference/chr14.fa

Using GATK jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar
Running:
    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -Xmx4G -jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar CreateSequenceDictionary -R ./reference/chr14.fa
INFO	2026-05-04 18:02:41	CreateSequenceDictionary	Output dictionary will be written in /Users/amira/Desktop/BIO392/exercises/day8/notebook/./reference/chr14.dict
18:02:41.602 INFO  NativeLibraryLoader - Loading libgkl_compression.dylib from jar:file:/opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar!/com/intel/gkl/native/libgkl_compression.dylib
[Mon May 04 18:02:42 CEST 2026] CreateSequenceDictionary --REFERENCE ./reference/chr14.fa --TRUNCATE_NAMES_AT_WHITESPACE true --NUM_SEQUENCES 2147483647 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGEN

In [9]:
!mkdir -p ./outputB/bqsr

In [10]:
!gatk --java-options "-Xmx4G" BaseRecalibrator \
        -R ./reference/chr14.fa \
        -I ./outputB/bwa/B_sort_markdup.bam \
        -O ./outputB/bqsr/B_sort_markdup_BQSR.report \
        --known-sites ./db/Homo_sapiens_assembly38.dbsnp138_roi.vcf.gz \
        --known-sites ./db/resources_broad_hg38_v0_Homo_sapiens_assembly38.known_indels.vcf.gz \
        --known-sites ./db/resources_broad_hg38_v0_Mills_and_1000G_gold_standard.indels.hg38.vcf.gz

Using GATK jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar
Running:
    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -Xmx4G -jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar BaseRecalibrator -R ./reference/chr14.fa -I ./outputB/bwa/B_sort_markdup.bam -O ./outputB/bqsr/B_sort_markdup_BQSR.report --known-sites ./db/Homo_sapiens_assembly38.dbsnp138_roi.vcf.gz --known-sites ./db/resources_broad_hg38_v0_Homo_sapiens_assembly38.known_indels.vcf.gz --known-sites ./db/resources_broad_hg38_v0_Mills_and_1000G_gold_standard.indels.hg38.vcf.gz
18:03:11.688 INFO  NativeLibraryLoader - Loading libgkl_compression.dylib from jar:file:/opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar!/com/intel/gkl/native/libgkl_compression.dylib
18:03:12.356 INFO  BaseRecalibrator - --------------

In [11]:
!gatk --java-options "-Xmx4G" ApplyBQSR \
        -I ./outputB/bwa/B_sort_markdup.bam \
        -O ./outputB/bqsr/B_sort_markdup_recal.bam \
        --bqsr-recal-file ./outputB/bqsr/B_sort_markdup_BQSR.report

Using GATK jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar
Running:
    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -Xmx4G -jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar ApplyBQSR -I ./outputB/bwa/B_sort_markdup.bam -O ./outputB/bqsr/B_sort_markdup_recal.bam --bqsr-recal-file ./outputB/bqsr/B_sort_markdup_BQSR.report
18:03:46.207 INFO  NativeLibraryLoader - Loading libgkl_compression.dylib from jar:file:/opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar!/com/intel/gkl/native/libgkl_compression.dylib
18:03:46.948 INFO  ApplyBQSR - ------------------------------------------------------------
18:03:46.955 INFO  ApplyBQSR - The Genome Analysis Toolkit (GATK) v4.6.2.0
18:03:46.956 INFO  ApplyBQSR - For support and documentation go to https://software.broadinstitute.o

In [12]:
!mkdir -p ./outputB/haplotypecaller
!mkdir -p ./tmp

In [13]:
!gatk --java-options "-Xmx8G" HaplotypeCaller \
    -R ./reference/chr14.fa \
    -I ./outputB/bqsr/B_sort_markdup_recal.bam \
    -O ./outputB/haplotypecaller/B.g.vcf \
    -ERC GVCF \
    --tmp-dir ./tmp

Using GATK jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar
Running:
    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -Xmx8G -jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar HaplotypeCaller -R ./reference/chr14.fa -I ./outputB/bqsr/B_sort_markdup_recal.bam -O ./outputB/haplotypecaller/B.g.vcf -ERC GVCF --tmp-dir ./tmp
18:04:25.922 INFO  NativeLibraryLoader - Loading libgkl_compression.dylib from jar:file:/opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar!/com/intel/gkl/native/libgkl_compression.dylib
18:04:26.580 INFO  HaplotypeCaller - ------------------------------------------------------------
18:04:26.587 INFO  HaplotypeCaller - The Genome Analysis Toolkit (GATK) v4.6.2.0
18:04:26.588 INFO  HaplotypeCaller - For support and documentation go to https://software.b

In [14]:
!ls ./outputB/haplotypecaller/*.g.vcf
!ls ./outputB/haplotypecaller/*.g.vcf | xargs -n1 basename
!ls ./outputB/haplotypecaller/*.g.vcf | xargs -n1 basename \
    | awk '{split($0,a,".");print a[1]}' > ./outputB/haplotypecaller/input_name.txt

!cat ./outputB/haplotypecaller/input_name.txt
!realpath ./outputB/haplotypecaller/*.g.vcf > ./outputB/haplotypecaller/input_path.txt
!cat ./outputB/haplotypecaller/input_path.txt

!paste ./outputB/haplotypecaller/input_name.txt \
    ./outputB/haplotypecaller/input_path.txt  > ./outputB/haplotypecaller/samples_map.txt

!cat ./outputB/haplotypecaller/samples_map.txt

./outputB/haplotypecaller/B.g.vcf
B.g.vcf
B
/Users/amira/Desktop/BIO392/exercises/day8/notebook/outputB/haplotypecaller/B.g.vcf
B	/Users/amira/Desktop/BIO392/exercises/day8/notebook/outputB/haplotypecaller/B.g.vcf


In [15]:
!gatk --java-options "-Xms8G -Xmx16G" GenomicsDBImport \
    --genomicsdb-workspace-path ./outputB/genomeDB \
    --batch-size 50 \
    -L chr14 \
    --sample-name-map ./outputB/haplotypecaller/samples_map.txt \
    --tmp-dir ./tmp \
    --reader-threads 2

Using GATK jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar
Running:
    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -Xms8G -Xmx16G -jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar GenomicsDBImport --genomicsdb-workspace-path ./outputB/genomeDB --batch-size 50 -L chr14 --sample-name-map ./outputB/haplotypecaller/samples_map.txt --tmp-dir ./tmp --reader-threads 2
18:08:11.416 INFO  NativeLibraryLoader - Loading libgkl_compression.dylib from jar:file:/opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar!/com/intel/gkl/native/libgkl_compression.dylib
18:08:12.340 INFO  GenomicsDBImport - ------------------------------------------------------------
18:08:12.349 INFO  GenomicsDBImport - The Genome Analysis Toolkit (GATK) v4.6.2.0
18:08:12.349 INFO  GenomicsDBImport - For s

In [16]:
!mkdir -p ./outputB/genotypegvcf
!gatk --java-options "-Xmx8G" GenotypeGVCFs \
    -R ./reference/chr14.fa \
    -V ./outputB/haplotypecaller/B.g.vcf \
    -O ./outputB/genotypegvcf/B.vcf \
    --tmp-dir ./tmp

!bcftools view -H ./outputB/genotypegvcf/B.vcf | head -10

Using GATK jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar
Running:
    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -Xmx8G -jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar GenotypeGVCFs -R ./reference/chr14.fa -V ./outputB/haplotypecaller/B.g.vcf -O ./outputB/genotypegvcf/B.vcf --tmp-dir ./tmp
18:08:35.594 INFO  NativeLibraryLoader - Loading libgkl_compression.dylib from jar:file:/opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar!/com/intel/gkl/native/libgkl_compression.dylib
18:08:36.341 INFO  GenotypeGVCFs - ------------------------------------------------------------
18:08:36.345 INFO  GenotypeGVCFs - The Genome Analysis Toolkit (GATK) v4.6.2.0
18:08:36.345 INFO  GenotypeGVCFs - For support and documentation go to https://software.broadinstitute.org/gatk/
18:08

In [17]:
!gatk --java-options "-Xmx8G" GenotypeGVCFs \
    -R ./reference/chr14.fa \
    -V gendb://${PWD}/outputB/genomeDB \
    -O ./outputB/genotypegvcf/combine.vcf \
    --tmp-dir ./tmp

Using GATK jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar
Running:
    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -Xmx8G -jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar GenotypeGVCFs -R ./reference/chr14.fa -V gendb:///Users/amira/Desktop/BIO392/exercises/day8/notebook/outputB/genomeDB -O ./outputB/genotypegvcf/combine.vcf --tmp-dir ./tmp
18:09:04.393 INFO  NativeLibraryLoader - Loading libgkl_compression.dylib from jar:file:/opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar!/com/intel/gkl/native/libgkl_compression.dylib
18:09:05.198 INFO  GenotypeGVCFs - ------------------------------------------------------------
18:09:05.203 INFO  GenotypeGVCFs - The Genome Analysis Toolkit (GATK) v4.6.2.0
18:09:05.204 INFO  GenotypeGVCFs - For support and documentation go t

In [18]:
!bcftools view -h ./outputB/genotypegvcf/combine.vcf | tail
!bcftools view -H ./outputB/genotypegvcf/combine.vcf | head -10

!mkdir -p ./outputB/vqsr

##INFO=<ID=RAW_MQandDP,Number=2,Type=Integer,Description="Raw data (sum of squared MQ and total depth) for improved RMS Mapping Quality calculation. Incompatible with deprecated RAW_MQ formulation.">
##INFO=<ID=ReadPosRankSum,Number=1,Type=Float,Description="Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias">
##INFO=<ID=SOR,Number=1,Type=Float,Description="Symmetric Odds Ratio of 2x2 contingency table to detect strand bias">
##contig=<ID=chr14,length=107043718>
##source=GenomicsDBImport
##source=GenotypeGVCFs
##source=HaplotypeCaller
##bcftools_viewVersion=1.23.1+htslib-1.23
##bcftools_viewCommand=view -h ./outputB/genotypegvcf/combine.vcf; Date=Mon May  4 18:09:35 2026
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	B
chr14	18601385	.	C	T	69.64	.	AC=1;AF=0.5;AN=2;BaseQRankSum=-1.123;DP=61;ExcessHet=0;FS=23.489;MLEAC=1;MLEAF=0.5;MQ=35.53;MQRankSum=-0.143;QD=1.22;ReadPosRankSum=-0.706;SOR=4.167	GT:AD:DP:GQ:PL	0/1:50,7:57:77:77,0,1515
chr14	18601430	.	A	C	632.64	.	AC=1;

In [19]:
!gatk --java-options "-Xmx8G" VariantRecalibrator \
    -R ./reference/chr14.fa \
    -V ./outputB/genotypegvcf/combine.vcf \
    --resource:hapmap,known=false,training=true,truth=true,prior=15.0 ./db/vqsr/resources_broad_hg38_v0_hapmap_3.3.hg38.vcf.gz  \
    --resource:omni,known=false,training=true,truth=false,prior=12.0 ./db/vqsr/resources_broad_hg38_v0_1000G_omni2.5.hg38.vcf.gz \
    --resource:1000G,known=false,training=true,truth=false,prior=10.0 ./db/vqsr/resources_broad_hg38_v0_1000G_phase1.snps.high_confidence.hg38_chr14roi.vcf.gz \
    --resource:dbsnp,known=true,training=false,truth=false,prior=2.0 ./db/Homo_sapiens_assembly38.dbsnp138_roi.vcf.gz \
    -an QD -an MQ -an MQRankSum -an ReadPosRankSum -an FS -an SOR  \
    -mode SNP \
    --max-gaussians 4 \
    -O ./outputB/vqsr/vqsr_snp.recal \
    --tranches-file ./outputB/vqsr/vqsr_snp.tranches

Using GATK jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar
Running:
    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -Xmx8G -jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar VariantRecalibrator -R ./reference/chr14.fa -V ./outputB/genotypegvcf/combine.vcf --resource:hapmap,known=false,training=true,truth=true,prior=15.0 ./db/vqsr/resources_broad_hg38_v0_hapmap_3.3.hg38.vcf.gz --resource:omni,known=false,training=true,truth=false,prior=12.0 ./db/vqsr/resources_broad_hg38_v0_1000G_omni2.5.hg38.vcf.gz --resource:1000G,known=false,training=true,truth=false,prior=10.0 ./db/vqsr/resources_broad_hg38_v0_1000G_phase1.snps.high_confidence.hg38_chr14roi.vcf.gz --resource:dbsnp,known=true,training=false,truth=false,prior=2.0 ./db/Homo_sapiens_assembly38.dbsnp138_roi.vcf.gz -an QD -an MQ -an MQRankSum -an Rea

In [20]:
!gatk --java-options "-Xmx8G" VariantRecalibrator \
    -R ./reference/chr14.fa \
    -V ./outputB/genotypegvcf/combine.vcf \
    --resource:mills,known=false,training=true,truth=true,prior=12.0 ./db/resources_broad_hg38_v0_Mills_and_1000G_gold_standard.indels.hg38.vcf.gz \
    --resource:dbsnp,known=true,training=false,truth=false,prior=2.0 ./db/Homo_sapiens_assembly38.dbsnp138_roi.vcf.gz \
    -an QD -an MQRankSum -an ReadPosRankSum -an FS -an SOR -an DP \
    -mode INDEL \
    --max-gaussians 1 \
    -O ./outputB/vqsr/vqsr_indel.recal \
    --tranches-file ./outputB/vqsr/vqsr_indel.tranches

Using GATK jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar
Running:
    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -Xmx8G -jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar VariantRecalibrator -R ./reference/chr14.fa -V ./outputB/genotypegvcf/combine.vcf --resource:mills,known=false,training=true,truth=true,prior=12.0 ./db/resources_broad_hg38_v0_Mills_and_1000G_gold_standard.indels.hg38.vcf.gz --resource:dbsnp,known=true,training=false,truth=false,prior=2.0 ./db/Homo_sapiens_assembly38.dbsnp138_roi.vcf.gz -an QD -an MQRankSum -an ReadPosRankSum -an FS -an SOR -an DP -mode INDEL --max-gaussians 1 -O ./outputB/vqsr/vqsr_indel.recal --tranches-file ./outputB/vqsr/vqsr_indel.tranches
18:10:09.520 INFO  NativeLibraryLoader - Loading libgkl_compression.dylib from jar:file:/opt/anaconda3/envs/human_env

In [29]:
!gatk --java-options "-Xmx8G" ApplyVQSR \
  -V ./outputB/genotypegvcf/combine.vcf \
  --recal-file ./outputB/vqsr/vqsr_snp.recal \
  -mode SNP \
  --tranches-file ./outputB/vqsr/vqsr_snp.tranches \
  --truth-sensitivity-filter-level 99.5 \
  --create-output-variant-index true \
  -O ./outputB/vqsr/combine_snp_recalibrated.vcf.gz

Using GATK jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar
Running:
    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -Xmx8G -jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar ApplyVQSR -V ./outputB/genotypegvcf/combine.vcf --recal-file ./outputB/vqsr/vqsr_snp.recal -mode SNP --tranches-file ./outputB/vqsr/vqsr_snp.tranches --truth-sensitivity-filter-level 99.5 --create-output-variant-index true -O ./outputB/vqsr/combine_snp_recalibrated.vcf.gz
15:48:22.550 INFO  NativeLibraryLoader - Loading libgkl_compression.dylib from jar:file:/opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar!/com/intel/gkl/native/libgkl_compression.dylib
15:48:23.524 INFO  ApplyVQSR - ------------------------------------------------------------
15:48:23.533 INFO  ApplyVQSR - The Genome Analysis 

In [21]:
!bcftools view -H ./outputB/vqsr/combine_snp_recalibrated.vcf.gz | head -10
!vcftools --gzvcf ./outputB/vqsr/combine_snp_recalibrated.vcf.gz \
    --recode \
    --recode-INFO-all \
    --remove-filtered-all \
    --out ./outputB/vqsr/combine_indel_snp_recalibrated

chr14	18601385	.	C	T	69.64	VQSRTrancheSNP99.90to100.00	AC=1;AF=0.5;AN=2;BaseQRankSum=-1.123;DP=61;ExcessHet=0;FS=23.489;MLEAC=1;MLEAF=0.5;MQ=35.53;MQRankSum=-0.143;QD=1.22;ReadPosRankSum=-0.706;SOR=4.167;VQSLOD=-78.74;culprit=FS	GT:AD:DP:GQ:PL	0/1:50,7:57:77:77,0,1515
chr14	18601430	.	A	C	632.64	VQSRTrancheSNP99.90to100.00	AC=1;AF=0.5;AN=2;BaseQRankSum=-5.495;DP=83;ExcessHet=0;FS=73.721;MLEAC=1;MLEAF=0.5;MQ=33.47;MQRankSum=0.693;QD=8.55;ReadPosRankSum=-0.132;SOR=6.517;VQSLOD=-752.7;culprit=FS	GT:AD:DP:GQ:PL	0/1:49,25:74:99:640,0,1470
chr14	18601553	.	T	A	3281.64	VQSRTrancheSNP99.90to100.00	AC=1;AF=0.5;AN=2;BaseQRankSum=-7.851;DP=284;ExcessHet=0;FS=30.232;MLEAC=1;MLEAF=0.5;MQ=38.36;MQRankSum=-4.714;QD=12.57;ReadPosRankSum=-2.149;SOR=2.105;VQSLOD=-132.4;culprit=FS	GT:AD:DP:GQ:PL	0/1:139,122:261:99:3289,0,4044
chr14	18601712	.	G	T	1177.64	PASS	AC=1;AF=0.5;AN=2;BaseQRankSum=-3.989;DP=94;ExcessHet=0;FS=3.04;MLEAC=1;MLEAF=0.5;MQ=44.67;MQRankSum=-8.252;QD=13.08;ReadPosRankSum=-1.088;SOR=2.144

Apply VQSR for INDEL variants.

In [22]:
!gatk --java-options "-Xmx8G" ApplyVQSR \
  -V ./outputB/vqsr/combine_snp_recalibrated.vcf.gz \
  -mode INDEL \
  --recal-file ./outputB/vqsr/vqsr_indel.recal \
  --tranches-file ./outputB/vqsr/vqsr_indel.tranches \
  --truth-sensitivity-filter-level 99.0 \
  --create-output-variant-index true \
  -O ./outputB/vqsr/combine_indel_snp_recalibrated.vcf.gz

Using GATK jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar
Running:
    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -Xmx8G -jar /opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar ApplyVQSR -V ./outputB/vqsr/combine_snp_recalibrated.vcf.gz -mode INDEL --recal-file ./outputB/vqsr/vqsr_indel.recal --tranches-file ./outputB/vqsr/vqsr_indel.tranches --truth-sensitivity-filter-level 99.0 --create-output-variant-index true -O ./outputB/vqsr/combine_indel_snp_recalibrated.vcf.gz
18:10:54.683 INFO  NativeLibraryLoader - Loading libgkl_compression.dylib from jar:file:/opt/anaconda3/envs/human_env/share/gatk4-4.6.2.0-0/gatk-package-4.6.2.0-local.jar!/com/intel/gkl/native/libgkl_compression.dylib
18:10:55.926 INFO  ApplyVQSR - ------------------------------------------------------------
18:10:55.941 INFO  ApplyVQS

In [24]:
!bcftools view -H ./outputB/vqsr/combine_indel_snp_recalibrated.vcf.gz | head -10

chr14	18601385	.	C	T	69.64	VQSRTrancheSNP99.90to100.00	AC=1;AF=0.5;AN=2;BaseQRankSum=-1.123;DP=61;ExcessHet=0;FS=23.489;MLEAC=1;MLEAF=0.5;MQ=35.53;MQRankSum=-0.143;QD=1.22;ReadPosRankSum=-0.706;SOR=4.167;VQSLOD=-78.74;culprit=FS	GT:AD:DP:GQ:PL	0/1:50,7:57:77:77,0,1515
chr14	18601430	.	A	C	632.64	VQSRTrancheSNP99.90to100.00	AC=1;AF=0.5;AN=2;BaseQRankSum=-5.495;DP=83;ExcessHet=0;FS=73.721;MLEAC=1;MLEAF=0.5;MQ=33.47;MQRankSum=0.693;QD=8.55;ReadPosRankSum=-0.132;SOR=6.517;VQSLOD=-752.7;culprit=FS	GT:AD:DP:GQ:PL	0/1:49,25:74:99:640,0,1470
chr14	18601553	.	T	A	3281.64	VQSRTrancheSNP99.90to100.00	AC=1;AF=0.5;AN=2;BaseQRankSum=-7.851;DP=284;ExcessHet=0;FS=30.232;MLEAC=1;MLEAF=0.5;MQ=38.36;MQRankSum=-4.714;QD=12.57;ReadPosRankSum=-2.149;SOR=2.105;VQSLOD=-132.4;culprit=FS	GT:AD:DP:GQ:PL	0/1:139,122:261:99:3289,0,4044
chr14	18601712	.	G	T	1177.64	PASS	AC=1;AF=0.5;AN=2;BaseQRankSum=-3.989;DP=94;ExcessHet=0;FS=3.04;MLEAC=1;MLEAF=0.5;MQ=44.67;MQRankSum=-8.252;QD=13.08;ReadPosRankSum=-1.088;SOR=2.144

Filter VCF — keep only PASS variants.

In [25]:
!vcftools --gzvcf ./outputB/vqsr/combine_indel_snp_recalibrated.vcf.gz \
    --recode \
    --recode-INFO-all \
    --remove-filtered-all \
    --out ./outputB/vqsr/combine_indel_snp_recalibrated_PASS


VCFtools - 0.1.17
(C) Adam Auton and Anthony Marcketta 2009

Parameters as interpreted:
	--gzvcf ./outputB/vqsr/combine_indel_snp_recalibrated.vcf.gz
	--recode-INFO-all
	--out ./outputB/vqsr/combine_indel_snp_recalibrated_PASS
	--recode
	--remove-filtered-all

Using zlib version: 1.3.1
After filtering, kept 1 out of 1 Individuals
Outputting VCF file...
After filtering, kept 2610 out of a possible 2688 Sites
Run Time = 0.00 seconds


Check unfiltered and filtered files (only PASS is kept).

In [26]:
!bcftools view -H ./outputB/vqsr/combine_indel_snp_recalibrated.vcf.gz | head -10
!bcftools view -H ./outputB/vqsr/combine_indel_snp_recalibrated_PASS.recode.vcf | head -10

chr14	18601385	.	C	T	69.64	VQSRTrancheSNP99.90to100.00	AC=1;AF=0.5;AN=2;BaseQRankSum=-1.123;DP=61;ExcessHet=0;FS=23.489;MLEAC=1;MLEAF=0.5;MQ=35.53;MQRankSum=-0.143;QD=1.22;ReadPosRankSum=-0.706;SOR=4.167;VQSLOD=-78.74;culprit=FS	GT:AD:DP:GQ:PL	0/1:50,7:57:77:77,0,1515
chr14	18601430	.	A	C	632.64	VQSRTrancheSNP99.90to100.00	AC=1;AF=0.5;AN=2;BaseQRankSum=-5.495;DP=83;ExcessHet=0;FS=73.721;MLEAC=1;MLEAF=0.5;MQ=33.47;MQRankSum=0.693;QD=8.55;ReadPosRankSum=-0.132;SOR=6.517;VQSLOD=-752.7;culprit=FS	GT:AD:DP:GQ:PL	0/1:49,25:74:99:640,0,1470
chr14	18601553	.	T	A	3281.64	VQSRTrancheSNP99.90to100.00	AC=1;AF=0.5;AN=2;BaseQRankSum=-7.851;DP=284;ExcessHet=0;FS=30.232;MLEAC=1;MLEAF=0.5;MQ=38.36;MQRankSum=-4.714;QD=12.57;ReadPosRankSum=-2.149;SOR=2.105;VQSLOD=-132.4;culprit=FS	GT:AD:DP:GQ:PL	0/1:139,122:261:99:3289,0,4044
chr14	18601712	.	G	T	1177.64	PASS	AC=1;AF=0.5;AN=2;BaseQRankSum=-3.989;DP=94;ExcessHet=0;FS=3.04;MLEAC=1;MLEAF=0.5;MQ=44.67;MQRankSum=-8.252;QD=13.08;ReadPosRankSum=-1.088;SOR=2.144

Compress and index the PASS-filtered VCF file.

In [27]:
!bgzip -c ./outputB/vqsr/combine_indel_snp_recalibrated_PASS.recode.vcf \
    > ./outputB/vqsr/combine_indel_snp_recalibrated_PASS.recode.vcf.gz
!tabix -p vcf ./outputB/vqsr/combine_indel_snp_recalibrated_PASS.recode.vcf.gz

## Structural variant calling

Using Delly on Sample B.

In [28]:
!mkdir -p ./outputB/delly
!delly call -o ./outputB/delly/B_dellySV.bcf \
    -g ./reference/chr14.fa \
    ./outputB/bqsr/B_sort_markdup_recal.bam

[2026-May-04 18:12:00] delly call -o ./outputB/delly/B_dellySV.bcf -g ./reference/chr14.fa ./outputB/bqsr/B_sort_markdup_recal.bam 
[2026-May-04 18:12:01] Paired-end and split-read scanning
[2026-May-04 18:12:01] Split-read clustering
[2026-May-04 18:12:01] Paired-end clustering
[2026-May-04 18:12:01] Split-read assembly
[2026-May-04 18:12:02] Generate REF and ALT probes
[2026-May-04 18:12:02] SV annotation
[2026-May-04 18:12:03] Genotyping
[2026-May-04 18:12:03] Sample statistics
Sample:A,ReadSize=101,Median=187,MAD=31,UniqueDiscordantPairs=176
[2026-May-04 18:12:03] Done.


In [29]:
!bcftools view -h ./outputB/delly/B_dellySV.bcf
!bcftools view -H ./outputB/delly/B_dellySV.bcf | head -10

##fileformat=VCFv4.2
##FILTER=<ID=PASS,Description="All filters passed">
##fileDate=20260504
##ALT=<ID=DEL,Description="Deletion">
##ALT=<ID=DUP,Description="Duplication">
##ALT=<ID=INV,Description="Inversion">
##ALT=<ID=BND,Description="Translocation">
##ALT=<ID=INS,Description="Insertion">
##FILTER=<ID=LowQual,Description="Poor quality and insufficient number of PEs and SRs.">
##INFO=<ID=CIEND,Number=2,Type=Integer,Description="PE confidence interval around END">
##INFO=<ID=CIPOS,Number=2,Type=Integer,Description="PE confidence interval around POS">
##INFO=<ID=CHR2,Number=1,Type=String,Description="Chromosome for POS2 coordinate in case of an inter-chromosomal translocation">
##INFO=<ID=POS2,Number=1,Type=Integer,Description="Genomic position for CHR2 in case of an inter-chromosomal translocation">
##INFO=<ID=END,Number=1,Type=Integer,Description="End position of the structural variant">
##INFO=<ID=PE,Number=1,Type=Integer,Description="Paired-end support of the structural variant">
#

Merge the SV candidates from Sample A and Sample B.

In [30]:
!delly merge -o ./outputB/delly/combine_dellySV.bcf \
    ./outputA/delly/A_dellySV.bcf \
    ./outputB/delly/B_dellySV.bcf

[2026-May-04 18:13:27] delly merge -o ./outputB/delly/combine_dellySV.bcf ./outputA/delly/A_dellySV.bcf ./outputB/delly/B_dellySV.bcf 
[2026-May-04 18:13:27] Reading input VCF/BCF files
[2026-May-04 18:13:27] Merging SVs
[2026-May-04 18:13:27] Filtering SVs
[2026-May-04 18:13:27] Done.
[2026-May-04 18:13:27] Reading input VCF/BCF files
[2026-May-04 18:13:27] Merging SVs
[2026-May-04 18:13:27] Filtering SVs
[2026-May-04 18:13:27] Done.
[2026-May-04 18:13:27] Reading input VCF/BCF files
[2026-May-04 18:13:27] Merging SVs
[2026-May-04 18:13:27] Filtering SVs
[2026-May-04 18:13:27] Done.
[2026-May-04 18:13:27] Reading input VCF/BCF files
[2026-May-04 18:13:27] Merging SVs
[2026-May-04 18:13:27] Filtering SVs
[2026-May-04 18:13:27] Done.
[2026-May-04 18:13:27] Reading input VCF/BCF files
[2026-May-04 18:13:27] Merging SVs
[2026-May-04 18:13:27] Filtering SVs
[2026-May-04 18:13:27] Done.
[2026-May-04 18:13:27] Reading input VCF/BCF files
[2026-May-04 18:13:27] Merging SVs
[2026-May-04 18:13:

In [31]:
!bcftools view -h ./outputB/delly/combine_dellySV.bcf
!bcftools view -H ./outputB/delly/combine_dellySV.bcf | head -10

##fileformat=VCFv4.2
##FILTER=<ID=PASS,Description="All filters passed">
##fileDate=20260504
##ALT=<ID=DEL,Description="Deletion">
##ALT=<ID=DUP,Description="Duplication">
##ALT=<ID=INV,Description="Inversion">
##ALT=<ID=BND,Description="Translocation">
##ALT=<ID=INS,Description="Insertion">
##FILTER=<ID=LowQual,Description="Poor quality and insufficient number of PEs and SRs.">
##INFO=<ID=CIEND,Number=2,Type=Integer,Description="PE confidence interval around END">
##INFO=<ID=CIPOS,Number=2,Type=Integer,Description="PE confidence interval around POS">
##INFO=<ID=CHR2,Number=1,Type=String,Description="Chromosome for POS2 coordinate in case of an inter-chromosomal translocation">
##INFO=<ID=POS2,Number=1,Type=Integer,Description="Genomic position for CHR2 in case of an inter-chromosomal translocation">
##INFO=<ID=END,Number=1,Type=Integer,Description="End position of the structural variant">
##INFO=<ID=PE,Number=1,Type=Integer,Description="Paired-end support of the structural variant">
#

Filter merged SVs: keep PASS calls and remove imprecise breakpoints.

In [32]:
!bcftools view -f PASS \
    -e 'INFO/IMPRECISE=1' \
    ./outputB/delly/combine_dellySV.bcf \
    -o ./outputB/delly/combine_dellySV_filter.bcf

In [ ]:
!bcftools view -H /Users/amira/Desktop/BIO392/exercises/day8/notebook/outputB/delly/combine_dellySV.bcf
!bcftools view -H /Users/amira/Desktop/BIO392/exercises/day8/notebook/outputB/delly/combine_dellySV_filter.bcf

chr14	104946713	DUP00000002	C	<DUP>	410	PASS	IMPRECISE;SVTYPE=DUP;SVMETHOD=EMBL.DELLYv1.7.3;END=104952186;PE=17;MAPQ=24;CT=5to3;CIPOS=-50,50;CIEND=-50,50
chr14	104948683	DEL00000001	T	<DEL>	744	PASS	IMPRECISE;SVTYPE=DEL;SVMETHOD=EMBL.DELLYv1.7.3;END=104949125;PE=40;MAPQ=22;CT=3to5;CIPOS=-50,50;CIEND=-50,50


Two structural variants were identified in the same region of chromosome 14 (around 104.9 Mb), both passing filters despite being imprecise calls: a ~5.5 kb duplication (DUP) supported by 17 paired-end reads, and a ~440 bp deletion (DEL) supported by 40 paired-end reads with a higher quality score of 744. The co-occurrence of a DUP and DEL in the same locus is notable — this could reflect a complex rearrangement, or the DUP may be a false positive arising from the abnormal read depth signature created by the nested deletion. Both calls have relatively low mapping quality (MAPQ 22–24) and wide confidence intervals (±50 bp on both breakpoints), so experimental validation would be warranted before drawing strong conclusions.